controller/program/src/instructions/liquidity/add.rs

```diff
@@ -0,0 +1,345 @@
+    // possible
+    if lst_amount_sol_value_after_fees.0 > *lst_amount_sol_value.0.start() {
+        return Err(Inf1CtlCustomProgErr(Inf1CtlErr::PoolWouldLoseSolValue).into());
+    }
+
+    let lp_token_supply = RawMint::of_acc_data(abr.get(*ix_prefix.lp_token_mint()).data())
+        .and_then(Mint::try_from_raw)
+        .map(|a| a.supply())
+        .ok_or(INVALID_ACCOUNT_DATA)?;
+
+    let add_liquidity_quote = quote_add_liq(AddLiqQuoteArgs {
+        amt: ix_args.amount,
+        lp_token_supply,
+        lp_mint: pool.lp_token_mint,
+        lp_protocol_fee_bps: pool.lp_protocol_fee_bps,
+        pool_total_sol_value: pool.total_sol_value,
+        inp_calc: lst_amount_sol_value,
+        pricing: lst_amount_sol_value_after_fees,
+        inp_mint: *abr.get(*ix_prefix.lst_mint()).key(),
+    })
+    .map_err(|e| ProgramError::from(AddLiqQuoteProgErr(e)))?;
+
+    // Step 6: lp_fees_sol_value = lp_tokens_sol_value - sol_value_to_add_after_fees
+    let to_reserves_lst_amount = match add_liquidity_quote
+        .0
+        .inp
+        .checked_sub(add_liquidity_quote.0.protocol_fee)
+    {
+        Some(reserves_fees) => reserves_fees,
+        None => return Err(Inf1CtlCustomProgErr(Inf1CtlErr::MathError).into()),
+    };
+
+    if to_reserves_lst_amount == 0 || add_liquidity_quote.0.out == 0 {
+        return Err(Inf1CtlCustomProgErr(Inf1CtlErr::ZeroValue).into());
+    }
+
+    if add_liquidity_quote.0.out < ix_args.min_out {
+        return Err(Inf1CtlCustomProgErr(Inf1CtlErr::SlippageToleranceExceeded).into());
+    }
+
+    let transfer_checked_accounts = NewTransferCheckedIxAccsBuilder::start()
+        .with_auth(*ix_prefix.signer())
+        .with_src(*ix_prefix.lst_acc())
+        .with_dst(*ix_prefix.pool_reserves())
+        .with_mint(*ix_prefix.lst_mint())
+        .build();
+
+    let lst_mint_acc_data = abr.get(*ix_prefix.lst_mint()).data();
+    let lst_mint_decimals = RawMint::of_acc_data(lst_mint_acc_data)
+        .and_then(Mint::try_from_raw)
+        .map(|a| a.decimals())
+        .ok_or(INVALID_ACCOUNT_DATA)?;
+
+    let token_prog = *abr.get(*ix_prefix.lst_mint()).owner();
+
+    let ix_data = TransferCheckedIxData::new(to_reserves_lst_amount, lst_mint_decimals);
+
+    // Transferring deposit fees to pool reserves
+    cpi.invoke_fwd(
+        abr,
+        &token_prog,
+        ix_data.as_buf(),
+        transfer_checked_accounts.0,
+    )?;
+
+    let transfer_checked_accounts = NewTransferCheckedIxAccsBuilder::start()
+        .with_auth(*ix_prefix.signer())
+        .with_src(*ix_prefix.lst_acc())
+        .with_dst(*ix_prefix.protocol_fee_accumulator())
+        .with_mint(*ix_prefix.lst_mint())
+        .build();
+
+    let ix_data = TransferCheckedIxData::new(add_liquidity_quote.0.protocol_fee, lst_mint_decimals);
+
+    // Transferring deposit fees to protrocol
+    cpi.invoke_fwd(
+        abr,
+        &token_prog,
+        ix_data.as_buf(),
+        transfer_checked_accounts.0,
+    )?;
```

Comment:
im not gonna tell you guys to try `array.map`-ing stuff anymore, will just cleanup afterwards